### PR TITLE
Function featurize

### PR DESCRIPTION
**Pull request by montoyjh**
A few changes to address #217.

Main change is in the addition of a `generate_feature_labels` method which takes column_ids as input.  Most Featurizers will not use column identity in generating labels, so the default is to return feature_labels, but the addition of this method in the base class allows developers of custom featurizers the ability to customize column names if desired without altering the implicit documentation role of the feature_labels method.

Also added a unittest to ensure FunctionFeaturizer works with MultipleFeaturizer.

### CODE REVIEW

**Comment by computron** [comment]
I don't really like this and I'm sure you don't either.

I don't think it's clear to me why this can't be incorporated into the existing ``fit()`` framework. The goal of ``fit()`` is just to set all internal variables that depend on your data choice.

Yes, ``fit()`` right now only takes in ``X`` and ``y`` data - but it also has the option to add ``fit_kwargs`` like col_ids or whatever you need for FunctionFeaturizer to set the internal label variables.  If the FunctionFeaturizer used fit() to get the right labels set internally, we'd be all set - is that correct?

Next, ``fit_featurize_dataframe()`` in BaseFeaturizer doesn't yet allow for throwing in ``**kwargs**`` to fit. But I think this is also fixable.

Are there other issues to using the existing fit() framework?

**Comment by montoyjh** [comment]
Yeah, I'm definitely not crazy about it, but I have two issues with using fit.  The first is that using `fit` to determine qualitative identity of output data seems semantically counterintuitive to me.  If I'm "fitting" something, I'd imagine I'm determining quantitative model parameters from data.  It's not obvious to me that I'd need to use it to determine qualitative labels for output.  

The other issue is that it makes any featurizer that relies on input column identity to describe feature labels stateful in a way that makes it prone to both soft and hard errors.  For example, I might generate labels that correspond to a certain set of columns, and then reuse the featurize_dataframe method with different columns.  If the number of columns is different, this probably throws an error telling the user something vague about how two numbers aren't the same 5 layers into a pandas traceback, but if they use the correct number of columns (but different columns), then featurize dataframe works fine but labels all of the columns incorrectly.

**Comment by ardunn** [comment]
@montoyjh 

regarding the 2nd issue:  I think some of the label mishap can be avoided by explicitly declaring what to do when a fittable featurizer is fit on a dataframe (generating one feature label set) and then featurizes another dataframe (assigning values based on a different potential feature label set). As far as I can tell, there are 4 scenarios where problems can occur:

1. Fitted set of labels has some labels not in the featurized set
2. Featurized set has some labels not in the fitted set
3. A combination of 1 and 2
4. No overlap between featurized set labels and fitted set labels

In BondFractions, for example, the fitted set of labels is always returned. 1 is taken care of by setting bond values in fitted set but not in featurized set to zero. 2 is taken care of by optionally approximating values for featurized labels to values for fitted labels, or dropping them otherwise. 3 is taken care of by using 1 and 2 in sequence. And 4 is taken care of by throwing an exception. Not saying a BondFractions-like implementation is ideal (or applicable to FunctionFeaturizer), but I haven't (yet) run into bad label assignments or vague 5-layer pandas tracebacks when fitting and featurizing on different sets.

**Comment by montoyjh** [comment]
Alright, I think this is moving in the prescribed direction.  What I really want to avoid is forcing the _user_ to use fit when using FunctionFeaturizer, because I don't want them to have to provide the same information (i. e. input column names) in multiple places.  In this implementation I put all of the logic for generating the labels in fit, and call it appropriately in both featurize_dataframe and fit_featurize_dataframe (because in this case it should be used in both).  The user could still use fit if they really wanted, but it wouldn't mess with the operation of featurize_dataframe.  I still need to resolve an issue in MultiFeaturizer compatibility related to the parallel use of featurization, which I'll tackle tomorrow.

**Comment by computron** [comment]
Btw @WardLT mentioned to me that he had some comments but is currently on travel. I was planning to wait for that before merging

Let me know however if you think this needs to be merged sooner rather than later

**Comment by montoyjh** [comment]
Sure, no reason to rush.  This is working now.  In `MultipleFeaturizer`, I've overriden featurize_dataframe to featurize the dataframe using each featurizer sequentially, which I think should allow any featurizer that overrides featurize_dataframe (as FunctionFeaturizer does) to be compatible.

**Comment by computron** [comment]
Ok great! I'll wait another week in case there are more reviews, otherwise I'll merge at that time and we can modify more after merging.

File: matminer/featurizers/function.py [hunk 1]
```diff
@@ -73,15 +75,53 @@ def exp_dict(self):
...
+        Sets the feature labels.  Not intended to be used by a user,
+        only intended to be invoked as part of featurize_dataframe
+
+        Args:
+            X ([str]): list of feature labels
+
+        Returns:
+            Set of feature labels corresponding to expressions
+        """
+        self._feature_labels = self.generate_string_expressions(X)
```
Comment (WardLT) [inline matminer/featurizers/function.py]: X should be a list of entries, and a not a list of column names.

I'd imagine the column names could be passed to this argument a few ways:
1. An optional keyword argument
2. The code detects if X is a dataframe and, if so, uses the column labels.

If the code does not detect label names, we could always make them up (e.g., "x0", "x1").

My reasoning for these suggestions are to make this featurizer follow a similar style to other TransformerMixin classes (which take data as input, not labels), and to allow people to  easily use this class without dataframes (I want to do that)

File: matminer/featurizers/base.py [hunk 2]
```diff
@@ -359,9 +359,21 @@ def __init__(self, featurizers):
     def featurize(self, *x):
         return np.hstack(np.squeeze(f.featurize(*x)) for f in self.featurizers)
 
+    def set_n_jobs(self, n_jobs):
+        for f in self.featurizers:
+            f.set_n_jobs(n_jobs)
```
Comment (WardLT) [inline matminer/featurizers/base.py]: Good call

File: matminer/featurizers/function.py [hunk 3]
```diff
@@ -73,15 +75,53 @@ def exp_dict(self):
...
+            list of functional outputs corresponding to input args
+        """
+        return list(self._exp_iter(*args, postprocess=self.postprocess))
+
+    def feature_labels(self):
+        """
+        Returns:
+            Set of feature labels corresponding to expressions
+        """
+        return self._feature_labels
```
Comment (WardLT) [inline matminer/featurizers/function.py]: Should we add in a "not fitted" error here?

File: matminer/featurizers/function.py [hunk 4]
```diff
@@ -91,46 +131,31 @@ def featurize_dataframe(self, df, col_id, ignore_errors=False,
...
-        return list(self._exp_iter(*args, postprocess=self.postprocess))
-
-    def feature_labels(self):
-        """
-        Returns:
-            Set of feature labels corresponding to expressions
+            updated dataframe based on featurizer fitted to that dataframe.
         """
-        return None
+        return self.featurize_dataframe(df, col_id, *args, **kwargs)
```
Comment (WardLT) [inline matminer/featurizers/function.py]: Do we need fit_featurize_dataframe if all it does is call featurize_dataframe?

File: matminer/featurizers/function.py [hunk 5]
```diff
@@ -73,15 +75,53 @@ def exp_dict(self):
...
+        """
+        Returns:
+            Set of feature labels corresponding to expressions
+        """
+        return self._feature_labels
+
+    def fit(self, X, y=None, **fit_kwargs):
+        """
+        Sets the feature labels.  Not intended to be used by a user,
+        only intended to be invoked as part of featurize_dataframe
```
Comment (WardLT) [inline matminer/featurizers/function.py]: This warning seems out of place, if you agree with my next comment.

File: matminer/featurizers/base.py [hunk 6]
```diff
@@ -359,9 +359,21 @@ def __init__(self, featurizers):
...
     def feature_labels(self):
         return sum([f.feature_labels() for f in self.featurizers], [])
 
+    def featurize_dataframe(self, df, col_id, ignore_errors=False,
+                            return_errors=False, inplace=True):
+        for f in self.featurizers:
+            df = f.featurize_dataframe(df, col_id, ignore_errors,
+                                       return_errors, inplace)
+            df[f.feature_labels()] = df[f.feature_labels()].applymap(np.squeeze)
+        return df
```
Comment (WardLT) [inline matminer/featurizers/base.py]: This change is going to wreck havoc with caching. If I understand multiprocessing correctly, each time featurize_dataframe completes the processes holding the cache will exit - thereby preventing data cached from one featurizer to be employed by another.

So, I think we need to be really sure we want this change to featurize data frame. Can you convince me? 

Fwiw, I'm not concerned FunctionFeaturize need work with MultiFeaturizer. I picture it being used in a Pipleline after primary features are computed but not in parallel (as assumed by MultiFeaturizer)

File: matminer/featurizers/tests/test_function.py [hunk 7]
```diff
@@ -82,6 +87,21 @@ def test_helper_functions(self):
...
+        self.assertEqual(len(test_combo_4), 18)
+
+    def test_multi_featurizer(self):
+        ff1 = FunctionFeaturizer(expressions=["x ** 2"])
+        ff2 = FunctionFeaturizer(expressions=["exp(x)", "1 / x"])
+        mf = MultipleFeaturizer([ff1, ff2])
+        new_df = mf.featurize_dataframe(self.test_df, ['a', 'b', 'c'],
+                                        inplace=False)
+        self.assertEqual(len(new_df), 11)
+
```
Comment (WardLT) [inline matminer/featurizers/tests/test_function.py]: Can we add a test for this class behaving well in a Pipeline? That's where I want to use it (and figure that may be a common use case)

File: matminer/featurizers/tests/test_function.py [hunk 8]
```diff
@@ -82,6 +87,21 @@ def test_helper_functions(self):
...
+                                                         combo_function=np.sum)
+        self.assertEqual(len(test_combo_4), 9)
+        test_combo_4 = generate_expressions_combinations(["1 / x", "x ** 2", "exp(x)"],
+                                                         combo_function=lambda x: x[1]-x[0])
+        self.assertEqual(len(test_combo_4), 18)
+
+    def test_multi_featurizer(self):
+        ff1 = FunctionFeaturizer(expressions=["x ** 2"])
+        ff2 = FunctionFeaturizer(expressions=["exp(x)", "1 / x"])
+        mf = MultipleFeaturizer([ff1, ff2])
```
Comment (WardLT) [inline matminer/featurizers/tests/test_function.py]: Ok. This might sell the need for FF to have MultiFeaturizer compatibility, but I still worry about giving up caching to save this. 

A few questions:

- Could we achieve the ability to make features based on functions of features that are functions of other features another way? Is this not what the depth argument provides? 
- Could we get it both ways by making a MultiFunctionFeaturizer? 
- Any other thoughts?

File: matminer/featurizers/function.py (hunk 1 above)
Comment (montoyjh) [reply matminer/featurizers/function.py]: Agree and prefer 2.  I tried 1 in another PR, and my initial approach to that was pretty unwieldy.  Latest commits have this.

File: matminer/featurizers/function.py (hunk 3 above)
Comment (montoyjh) [reply matminer/featurizers/function.py]: Added, tested.

File: matminer/featurizers/tests/test_function.py (hunk 7 above)
Comment (montoyjh) [reply matminer/featurizers/tests/test_function.py]: I don't know what a Pipeline is.  Someone who will use it that way should probably add that test (and any of the necessary functionality).

File: matminer/featurizers/base.py (hunk 6 above)
Comment (montoyjh) [reply matminer/featurizers/base.py]: Hmm... this seems like an important problem but I don't know how I'd fix it off of the top of my head.  Any featurizer that overrides featurize_dataframe will end up needing something like this, though (I don't know if any of the other featurizers do this).

To your last point, my main reason for trying to make this work is more practical, which is to say that someone actually tried to use it that way, it didn't work, and they e-mailed us to ask why. I think new users encountering a lack of interoperability of the code is bad for its perceived robustness, but I also think you're right that the more important supported use case should be in sequential applications of featurization, rather than parallel.

Comment (montoyjh) [reply matminer/featurizers/base.py]: Also, can you give me an example where this causes the problem you mention in a significant way?  I'm confident there is one, I just don't know the ins and outs of featurizers that use caching well enough to come up with an example on my own.

File: matminer/featurizers/tests/test_function.py (hunk 8 above)
Comment (montoyjh) [reply matminer/featurizers/tests/test_function.py]: To (1) - this doesn't achieve that, this test isn't really the intended use case, it's just to ensure that it works with MultiFeaturizer.  Thus, I don't think (2) is necessary.

One thing we could do is to just to create two modes of MultiFeaturizer operation, one "fast" mode that uses the prior method if FunctionFeaturizer (or other featurizers that override featurize_dataframe) is not present, and one "slow" mode that warns the user if a MultiFeaturizer-caching-incompatible featurizer is present.  This way, it still works if you use FunctionFeaturizer (and shouldn't be too slow if they're not using a bunch of features that are accelerated by caching) but issues a relevant warning that users can pay attention to if they're concerned about efficiency.

Comment (computron) [reply matminer/featurizers/tests/test_function.py]: I haven't played around too much with sklearn Pipeline, but perhaps someone can explain what MultiFeaturizer achieves that Pipeline can't. i.e., should we just encourage everyone to do multi-featurization in Pipelines?

Comment (WardLT) [reply matminer/featurizers/tests/test_function.py]: Scikit-learn's Pipeline tool is used to construct models that are composed of multiple, sequential operations. A simple example would be a model that first converts a pmg Composition into a feature vector and then trains a linear regression model . The distinguishing feature of a Pipeline is that the outputs from one step are used as the inputs into the next one. In contrast, MultiFeaturizer is provides the same inputs to all of the suboperations.

An example that uses MultiFeaturizer, FunctionFeaturizer, and Pipeline together is:

```python
model = Pipeline([
  ('featurize', MultiFeaturizer([ElementProperty(), IonProperty()])),
  ('expand', FunctionFeaturizer()),
  ('model', Lasso())
])
model.fit([Composition('Al')], [0])
```

This Pipeline will first call MultiFeaturizer transform the list of Compositions into a feature array, then use FunctionFeaturizer to expand those primary features, and then use the expanded list of features to fit a Lasso model. 

Sorry for the long reply, but does this help clarify how all of these tools could fit together?

File: matminer/featurizers/base.py (hunk 6 above)
Comment (WardLT) [reply matminer/featurizers/base.py]: The Voronoi notebook in our [examples repo](https://github.com/hackingmaterials/matminer_examples/blob/master/notebooks/machine-learning/voronoi-ward-prb-2017.ipynb) uses caching with MultiFeaturizer.

File: matminer/featurizers/tests/test_function.py (hunk 8 above)
Comment (computron) [reply matminer/featurizers/tests/test_function.py]: I see, but why can't someone do it this way?

```
model = Pipeline([
  ('featurize1', ElementProperty()),
  ('featurize2', IonProperty()),
  ...
  ('expand', FunctionFeaturizer()),
  ('model', Lasso())
])
```

If the above is valid, then the main use case I see is that MultiFeaturizer is just more intuitive than sklearn Pipelines. Technically, MultiFeaturizer could also parallelize across featurizers (instead of across data within featurizers), but I don't think that's being exploited yet.

But given that Pipelines do more, my question is whether we should encourage people to bite the bullet and learn Pipelines - e.g., by giving them an easy example notebook to work from. Since this is a more complex discussion I'll add it as a separate issue.

**Comment by computron** [comment]
@montoyjh @WardLT 

I'm a bit lost on the status of this PR.

Is this ready to merge?

If not, what steps are remaining?

File: matminer/featurizers/tests/test_function.py (hunk 8 above)
Comment (computron) [reply matminer/featurizers/tests/test_function.py]: Or are you saying that the above *doesn't* work since ElementProperty doesn't just return an array of Compositions (rather a matrix of Composition + other features), so the IonProperty() step would fail?

Comment (WardLT) [reply matminer/featurizers/tests/test_function.py]: Yes, the above code you just provided would not work because ElementProperty returns a list of features and not a list of Compositions, as needed by IonProperty.

**Comment by WardLT** [comment]
I don't think this is ready to merge. 

The one critical issue, for me, is that overloading `featurize_dataframe` kills our caching. If we could fix that, then I'm satisfied. The other things we can address later.

**Comment by montoyjh** [comment]
We need to decide on how we want multifeaturizer to work, we have three options currently proposed

(1) - What's in this pull requests changes `MultiFeaturizer.featurize_dataframe` to run the featurizers sequentially, rather than in aggregate using the `MultiFeaturizer.featurize` method and the inherited `BaseFeaturizer.featurizer_dataframe`.  FunctionFeaturizer is compatible with MultiFeaturizer (and presumably any Featurizer that overrides featurize_dataframe is also), but we are left with potential issues with caching.
(2) - I can revert the changes so we keep what's been changed in FunctionFeaturizer (which I think is an improvement, even if #217 isn't addressed).  FunctionFeaturizer will not be compatible with MultiFeaturizer.
(3) - We could adopt a default mode for multifeaturizer if none of the included featurizers override featurize_dataframe which would do everything in parallel, and then a sequential mode if any of the included featurizers override featurize_dataframe, issuing a warning that performance may be hindered by the sequential operation.

**Comment by montoyjh** [comment]
@WardLT How do you feel about option 3 above?  I'm not crazy about making MultiFeaturizer more complicated, but it presumably wouldn't change anything about how MatMiner performs now.

**Comment by WardLT** [comment]
Options 3 sounds OK to me. I share your concern about making MultiFeaturizer complicated, but it would allow us to have the best of options 1 and 2 (i.e., caching and MF compatibility). 

How do you detect if a function has been overloaded?

**Comment by WardLT** [comment]
Also, in the interest of closing this pull request, let's tackle the test for Pipeline compatibility in another PR. I might be able to work on that early next week.

**Comment by montoyjh** [comment]
Okay, I modified such that option 3 is implemented as indicated above (with override detection in the class dict), let me know how that looks.

**Comment by WardLT** [comment]
This looks good to me. Thanks @montoyjh !
